pkg/controllers/node/health/controller.go

```diff
@@ -0,0 +1,121 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+)
+
+// Controller for the resource
+type Controller struct {
+	clock         clock.Clock
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider, clock clock.Clock) *Controller {
+	return &Controller{
+		clock:         clock,
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("node.health").
+		For(&corev1.Node{}).
+		Complete(reconcile.AsReconciler(m.GetClient(), c))
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *corev1.Node) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "node.health")
+	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("node", klog.KRef(node.Namespace, node.Name)))
+
+	// Validate that the node is owned by us and is not being deleted
+	if !controllerutil.ContainsFinalizer(node, v1.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+
+	// If find if a node is unhealthy
+	healthCondition, repairNode := lo.Find(c.cloudProvider.RepairPolicies(), func(policy cloudprovider.RepairPolicy) bool {
+		unhealthyNodeCondition := nodeutils.GetCondition(node, policy.ConditionType)
+		if unhealthyNodeCondition.Status != policy.ConditionStatus {
+			return false
+		}
+		// If the Node is unhealthy, but has not reached it's full toleration disruption
+		terminationTime := unhealthyNodeCondition.LastTransitionTime.Add(policy.TolerationDuration)
+		return c.clock.Now().After(terminationTime)
+	})
+	if !repairNode {
+		return reconcile.Result{}, nil
+	}
+
+	// For unhealthy past the tolerationDisruption window we can forcefully terminate the node
+	if err := c.annotateTerminationGracePeriod(ctx, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("annotated termination grace period on nodeclaim, %w", err)
```

Comment:
client.IgnoreNotFound?

Reply:
I'm little confused here. Did I miss understand this comment? https://github.com/kubernetes-sigs/karpenter/pull/1793#discussion_r1843348779

Reply:
I get what you mean the previous comment. Discussed ofline